earth_enterprise/src/fusion/autoingest/sysman/AssetOperation.cpp

```diff
@@ -87,9 +87,57 @@ void HandleTaskProgress(const TaskProgressMsg & msg) {
   }
 }
 
-void UpdateWaitingAssets(const SharedString & ref, AssetDefs::State oldState) {
-  if (MiscConfig::Instance().GraphOperations) {
-    auto version = AssetVersion::storageManager().GetMutable(ref);
-    updater.UpdateWaitingAssets(version, oldState);
+void HandleTaskDone(const TaskDoneMsg & msg, bool graphOps) {
+  if (graphOps) {
+    auto version = assetOpStorageManager->GetMutable(msg.verref);
+    if (version && version->taskid == msg.taskid) {
+      version->beginTime = msg.beginTime;
+      version->progressTime = msg.endTime;
+      version->endTime = msg.endTime;
+      if (msg.success) {
+        version->ResetOutFiles(msg.outfiles);
+        stateUpdater->SetSucceeded(version);
+      }
+      else {
+        stateUpdater->SetFailed(version);
+      }
+    }
+    else if (!version) {
+      notify(NFY_WARN, "Could not load %s to mark done", msg.verref.c_str());
+    }
+  }
+  else {
+    AssetVersionD ver(msg.verref);
+    if (ver && ver->taskid == msg.taskid) {
+      MutableAssetVersionD(msg.verref)->HandleTaskDone(msg);
+    }
+    else if (!ver) {
+      notify(NFY_WARN, "Could not load %s to mark done", msg.verref.c_str());
+    }
+  }
+}
+
+// This function ensures that the State Updater stays in sync with any state
+// changes that happen in the legacy code.
+void HandleExternalStateChange(
+    const SharedString & ref,
+    AssetDefs::State oldState,
+    uint32 numInputsWaitingFor,
+    uint32 numChildrenWaitingFor,
+    bool graphOps) {
+  if (graphOps) {
+    auto version = assetOpStorageManager->Get(ref);
+    if (version) {
+      // Update the lists of waiting assets
+      stateUpdater->UpdateWaitingAssets(version, oldState, {numInputsWaitingFor, numChildrenWaitingFor});
```

Comment:
I think UpdateWaitingAssets may get called twice for a single update. Once here in response to a version's `OnStateChange` getting called, and once before this in `StateUpdater::RunStateChangeHandlers`. Is that right? I don't suppose it would cause an error, but could be extra work.

Reply:
Good catch; you're right about that. I moved the call in AssetVersionD from SetState to OnStateChange mostly because the child waiting count is a member of CompositeAssetVersionImplD, and SetState is defined in a parent class, so it doesn't have access to that variable. It's going to take some thinking to decide on the best way to handle that.

Reply:
Fixed.